First implementation of test tier tagging support

Jurisak opened the pull request:
Added resultDB query plugin. It is not functional now.
We need to wait till resultdb listener update.

Commit:
First implementation of test tier tagging support

Added resultDB query plugin. It is not functional now.
We need to wait till resultdb listener update.

Commit:
Query resultdb for jobs if only nvr and ci_tier was given

Commit:
Querying resultdb for job names. Uase case only with ci_tier and nvr provided

gsr-shanks commented:
Please add docstrings.

Jurisak commented:
I'll add them and also with tests. @gsr-shanks did you enable travis for this project?

gsr-shanks commented:
No, I did not enable travis.

Commit:
Method refactorization

Commit:
New way of resultdb query with added tests

Commit:
Merge branch 'master' of github.com:Jurisak/metamorph into test-tier-status

Commit:
Added docstrings and intalation of requests in travis ci

Commit:
Skiped one test because of RH sites

gsr-shanks commented:
Please add docs as well.

Commit:
Test tier status plugin documentation

Commit:
Added active waiting till results won't be in resultsDB

Commit:
Code refactorization

gsr-shanks commented on plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,211 @@
+        response = requests.get(url, params=url_options)
+        if response.status_code >= 300:
+            logging.error("ERROR: Unable to access resultsdb site.")
+            raise ResultsDBApiException("ERROR: Unable to access resultsdb site.")
```
Can we retry this for 3 times with sleep of 120 seconds before raising ResultsDBApiException ?

Jurisak replied:
Sure, I'll do.

Commit:
Added active waiting in query_resultsdb method

Commit:
Added better nvr parsing

Now user can choose from ci-message, environmental variable or
pass it as is.

gsr-shanks commented:
Nice. Getting NVR from ci-message should be the default. Thanks.

gsr-shanks commented:
`creating /usr/lib/python2.7/site-packages/plugins
copying build/lib/plugins/morph_messagehub.py -> /usr/lib/python2.7/site-packages/plugins
copying build/lib/plugins/morph_resultsdb.py -> /usr/lib/python2.7/site-packages/plugins
creating /usr/lib/python2.7/site-packages/plugins/library
copying build/lib/plugins/library/messagehub.py -> /usr/lib/python2.7/site-packages/plugins/library
copying build/lib/plugins/__init__.py -> /usr/lib/python2.7/site-packages/plugins
creating /usr/lib/python2.7/site-packages/lib
copying build/lib/lib/logging_conf.py -> /usr/lib/python2.7/site-packages/lib
copying build/lib/lib/__init__.py -> /usr/lib/python2.7/site-packages/lib`

packages are getting installed in site-packages basedir instead of site-packages/metamorph/

Jurisak commented:
Fixed in #6. Please merge that pull request as soon as possible.

gsr-shanks commented:
_plugins/morph_resultsdb.py_ should be _metamorph/plugins/morph_resultsdb.py_

Jurisak commented:
@gsr-shanks Sure i know. I've updated #6 pull request. Please merge it

Commit:
Merge branch 'master' of github.com:Jurisak/metamorph into test-tier-status

gsr-shanks commented on plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+import os
+import requests
+
+from lib.logging_conf import setup_logging
```
should be `from metamorph.lib.logging_conf`

Jurisak replied:
Yes but also no. In current state no metamoprh folder exists so it's correct. I'll solve it in a minute.

Commit:
Code refactorization due new project structural changes

Jurisak commented:
I would like to also mention that this plugin is not working yet. You can only try it on resultsdb devel stage because resultsdb listener changes were not moved into production, yet.

gsr-shanks commented:
Ok, could you please email me the command that I can try with resultsdb devel stage ?

gsr-shanks commented:
@robnester-rh could you please take a look

Jurisak commented:
@gsr-shanks I'll create ansible module after resultsdb updater changes will be merged in production.

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+        elif response.status_code >= 400:
+            logging.error("ERROR: Unable to access resultsdb site.")
+            raise ResultsDBApiException("ERROR: Unable to access resultsdb site.")
+        return response.json()
```
You could consider reworking this method to use the requests.raise_for_status() method to catch issues with communicating with the API.
The following seems to be a bit more compact to achieve the same result.

```
    logging.debug('Running resultsdb API query with this url: {0} and options {1}'.format(url, url_options))
    try:
        response = requests.get(url, params=url_options)
        response.raise_for_status()
        return response.json()
    except requests.HTTPError as e:
        if attempt < 3:
            attempt += 1
            time.sleep(120)
            self.query_resultsdb(url, url_options, attempt)
        else:
            logging.error("ERROR: Unable to access resultsdb site.")
            logging.error("ERROR: {0}".format(e.args)) # If you want to pass the exact error.
```

Jurisak replied:
Wow this looks really good. I didn't know about raise_for_status() method.
Thanks for this suggestion.

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        formatted_output = {}
+        for single_job in resultsdb_data:
+            formatted_output[single_job['job_names']] += single_job
```
From your docstring I think you're wanting to create a dictionary of the following structure:
```
formatted_output['job_names']<key = some job name> = <data from resultsdb_data>
```

If this is the case, then this method should be updated as such:

```
formatted_output = dict(job_names=dict())
for single_job in resultsdb_data:
    formatted_output['job_names'][single_job] = resultsdb_data[single_job] # if resultsdb_data is a dictionary of data, which I'm assuming it is from your existing structure.
return formatted_output
```

Jurisak replied:
resultsdb_data is list of queried resultsdb data (dictionaries). Therefore it can't be that way.

robnester-rh replied:
If resultsdb_data is a list of dictionaries, it still wouldn't work the way you've written it here I think.
If I understand you correctly, the following should be analogous to what you'd have:

```
resultsdb_data = [ {'job_names': 'a'}, {'job_names': 'b'}, {'job_names': 'c'}, {'job_names': 'd'} ]
formatted_output = {}
for single_job in resultsdb_data:
    formatted_output[single_job['job_names']] += single_job
print(formatted_output)
```
And this doesn't work. 

Can you clarify what your formatted_output dictionary structure should look like?

Jurisak replied:
Yes I know how you meant it but resultsdb_data looks like this:
`resultsdb_data [ { "data": { .. }, "job_names": "...", "id": [], "outcome": "PASS" }, { "data": ..., "job_names": ... }]`

For output dictionary structure see: [https://url.corp.redhat.com/ed75ec6]

robnester-rh replied:
Ah, ok. I see now. I wasn't assuming duplicate dict structures. Thanks for the clarification!

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+
+    def format_job_name_result(self, job_name_result):
+        """
+        Format's single job name data into dictionary
```
nit: correct to "Formats"

Jurisak replied:
Good catch. Thanks

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+
+        :returns -- Queried data
+        """
+        logging.debug('Running resultsbd API query with this url: {0} and options {1}'.format(url, url_options))
```
nit: correct typo  from 'resultsbd' to 'resultsdb'

Jurisak replied:
Thanks

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+    """
+    TIMEOUT_LIMIT = 7200  # Wait 2 hours maximally
+    RESULTSDB_LIMITER = 10
+    MINUTE = 60  # 60 seconds
```
I don't really think it's necessary to define MINUTE as a class constant. This seems to lead to having to do math on its value instead of just passing in given integers. Is it really any more clear to say sleep(self.MINUTE + self.MINUTE) instead of saying sleep(120) ?

Jurisak replied:
I generally don't like having magic numbers in my code.
@gsr-shanks  What do you think?

gsr-shanks replied:
@Jurisak commented below.

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+        response = requests.get(url, params=url_options)
+        if 300 <= response.status_code <= 400:
+            if attempt > 0:
+                time.sleep(self.MINUTE + self.MINUTE)  # Sleeping for two minutes
```
See above comment regarding defining this as a class constant.

Jurisak replied:
We can discuss this in above comment.

gsr-shanks replied:
I don't think we need to do math in this case. sleep(self.MINUTE) is good enough time to retry if the first try is unsuccessful.

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+        self.tier_tag = True
+        self.url_options = {'CI_tier': test_tier, 'item': component_nvr}
+
+    def get_test_tier_metadata(self):
```
The get_test_tier_metadata method appears to be an unused class method, is this for some future usage?

I'm curious also in that its name seems to suggest that it would return test tier metadata, but it's calling get_resultsdb_data, which doesn't return test tier metadata explicitly it seems?

Jurisak replied:
Good catch. Yes I want to use it in the main function.
I'll think about a new name.

robnester-rh commented on metamorph/plugins/morph_resultsdb.py:
```diff
@@ -0,0 +1,225 @@
+import requests
+
+from metamorph.lib.logging_conf import setup_logging
+
```
I don't know if it's a RH standard, but I generally separate stdlib imports, 3rd party imports and then project local imports by a blank line, such as the following:
```
import argparse
import logging
import logging.config
import json
import time
import os

import requests

from metamorph.lib.logging_conf import setup_logging
```

Jurisak replied:
Thanks that's a good attitude. I'll fix it.

robnester-rh reviewed (changes requested):
I think you have one "need to fix" in morph_resultsdb.py around line 100 regarding creating your dictionary of data that you're returning in the setup_output_data() method

Additionally, I would encourage looking at the comment regarding usage of .raise_for_status around line 47 in  the query_resultsdb() method in the same file.

Otherwise, a few nits / typos to fix.

Commit:
Code refactorization after code review

Commit:
Little code refactorization after code review

sleep.(MINUTE + MINUTE) was refactored to sleep.(MINUTE)
Also one logging information was added

Commit:
Fixed problem in setup_output_data method

Also updated saving metamorph.json

Commit:
Resultsdb ansible module implementation

Includes import separation in messagehub module

Commit:
Documentation update due ansible module creation

Commit:
Added path to certificate bundle

Commit:
Erased duplicity results and fixed tests

Commit:
Test input and output modifications

Commit:
ResultsDB ansible module update

+ Updated documentation due these changes

gsr-shanks commented:
There is still one change requested in this PR.

gsr-shanks commented:
@robnester-rh Looks like the changes requested are all in. Could you please merge this if all good ?

robnester-rh commented on tests/sources/resultsdb_output.json:
```diff
@@ -18,7 +18,7 @@
       "note": null,
       "outcome": "PASSED",
-      "ref_url": "http://someurl/864/console",
+      "ref_url": "http://someursl/864/console",
```
fix typo "someursl"

robnester-rh commented on tests/sources/resultsdb_output_result.json:
```diff
@@ -1 +1 @@
-{"result": {"tier": {"ci_tier": "", "tier_tag": true, "job_name": [{"setup-2.8.71-5.el7_1": [{"build_status": "PASSED", "build_url": "http://someurl/864", "build_number": "864"}, {"build_status": "PASSED", "build_url": "http://someurl/864", "build_number": "864"}]}], "nvr": ""}}}
\ No newline at end of file
+{"results": {"tier": {"ci_tier": "", "job_name": [{"setup-2.8.71-5.el7_1": [{"build_number": "864", "build_status": "PASSED", "build_url": "http://someursl/864"}, {"build_number": "864", "build_status": "PASSED", "build_url": "http://someurl/864"}]}], "nvr": "", "tier_tag": true}}}
```
fix typo "someursl"

robnester-rh reviewed (commented):
fix typo "someursl" where mentioned.

Commit:
Fixed typos in sources json files

Jurisak commented:
@robnester-rh Updated
